### PR TITLE
Warmup model before vLLM first time returns healthy signal

### PR DESCRIPTION
**Pull request by nostojicTT**
## Purpose
- Enabling warmup runs for prefill and decode in order to capture traces before the vLLM returns the healthy signal
- Before, this was done as a part of the first request, so the first user would experience bad performance

- `tt-metal` PR which is also part of the feature - https://github.com/tenstorrent/tt-metal/pull/33143

- link to issue that this PR resolves - https://github.com/tenstorrent/vllm/issues/220

- latest `vllm-nightly` run - https://github.com/tenstorrent/tt-metal/actions/runs/19941353422

### CODE REVIEW

File: vllm/v1/worker/tt_model_runner.py [hunk 1]
```diff
@@ -924,3 +924,12 @@ def generate_runner_output(self, sampled_token_ids: torch.Tensor):
...
         )
+
+    def warmup_model(self) -> None:
+        logger.info("Compile run for prefill started")
+        self.model.warmup_model_prefill(self.kv_caches, self.trace_mode)
+        logger.info("Compile run for prefill finished")
+
+        logger.info("Compile run for decode started")
+        self.model.warmup_model_decode(self.kv_caches, self.trace_mode, self.cache_config.num_gpu_blocks)
+        logger.info("Compile run for decode finished")
```
Comment (skhorasganiTT) [inline vllm/v1/worker/tt_model_runner.py]: "Compile run" -> "Warmup run". Same comment for V0.

File: vllm/worker/tt_worker.py [hunk 2]
```diff
@@ -249,6 +251,15 @@ def _init_cache_engine(self):
...
         self.tt_cache = self.cache_engine.tt_cache
 
+    def warmup_model(self) -> None:
+        logger.info("Compile run for prefill started")
+        self.model_runner.model.warmup_model_prefill(self.kv_cache, self.trace_mode)
+        logger.info("Compile run for prefill finished")
+        
+        logger.info("Compile run for decode started")
+        self.model_runner.model.warmup_model_decode(self.kv_cache, self.trace_mode, self.cache_config.num_gpu_blocks)
+        logger.info("Compile run for decode finished")
```
Comment (skhorasganiTT) [inline vllm/worker/tt_worker.py]: This function should be in model runner, same as V1.

File: vllm/worker/tt_worker.py [hunk 3]
```diff
@@ -237,6 +237,8 @@ def initialize_cache(
 
         self._init_cache_engine()
 
+        self.warmup_model()
```
Comment (skhorasganiTT) [inline vllm/worker/tt_worker.py]: Please call model_runner.warmup_model here directly

File: vllm/v1/worker/tt_model_runner.py [hunk 4]
```diff
@@ -28,6 +28,12 @@
 
 logger = init_logger(__name__)
 
+def create_warmup_decode_input_parameters(max_batch_size, num_gpu_blocks, sample_on_device_mode):
+    tokens = torch.zeros(max_batch_size, 1, dtype=torch.int32)
+    start_pos = torch.zeros(max_batch_size, dtype=torch.int32)
+    page_table = torch.zeros(max_batch_size, num_gpu_blocks, dtype=torch.int32)
+    sampling_params = TTSamplingParams(temperature=0.0, top_k=1, top_p=0.0) if sample_on_device_mode else None
+    return tokens, start_pos, page_table, sampling_params
```
Comment (skhorasganiTT) [inline vllm/v1/worker/tt_model_runner.py]: Please remove this and import from v0 model runner. Once v0 is deleted we can move that function to V1.

File: vllm/worker/tt_model_runner.py [hunk 5]
```diff
@@ -1119,3 +1125,14 @@ def _get_next_token_ids_from_sampler_output(
...
+    def warmup_model(self, kv_cache, trace_mode) -> None:
+        logger.info("Warmup run for prefill started")
+        self.model.warmup_model_prefill(kv_cache, trace_mode)
+        logger.info("Warmup run for prefill finished")
+        
+        tokens, start_pos, page_table, sampling_params = create_warmup_decode_input_parameters(self.scheduler_config.max_num_seqs, self.cache_config.num_gpu_blocks, self.sample_on_device_mode)
+        logger.info(f"Warmup run for decode with tokens: {tokens.shape}, start_pos: {start_pos.shape}, page_table: {page_table.shape}")
+        logger.info("Warmup run for decode started")
+        self.model.decode_forward(tokens, start_pos, page_table, kv_cache, trace_mode, sampling_params=sampling_params)
+        logger.info("Warmup run for decode finished")
```
Comment (skhorasganiTT) [inline vllm/worker/tt_model_runner.py]: If this code is the same for V1 maybe it should be moved into another util function above and called within this warmup_model function and the one in V1.

File: vllm/worker/tt_model_runner.py [hunk 6]
```diff
@@ -1119,3 +1125,14 @@ def _get_next_token_ids_from_sampler_output(
             for seq_output in seq_group_output.samples:
                 next_token_ids.append(seq_output.output_token)
         return torch.tensor(next_token_ids, dtype=torch.int32, device="cpu")
+
+    def warmup_model(self, kv_cache, trace_mode) -> None:
+        logger.info("Warmup run for prefill started")
+        self.model.warmup_model_prefill(kv_cache, trace_mode)
```
Comment (skhorasganiTT) [inline vllm/worker/tt_model_runner.py]: We actually don't use the trace mode arg for prefill forward today (since previously we didn't support traced prefill), that needs to be fixed as well in both v1 and v0 model runners.

File: vllm/worker/tt_model_runner.py [hunk 7]
```diff
@@ -1119,3 +1125,14 @@ def _get_next_token_ids_from_sampler_output(
...
+        logger.info("Warmup run for prefill finished")
+        
+        tokens, start_pos, page_table, sampling_params = create_warmup_decode_input_parameters(self.scheduler_config.max_num_seqs, self.cache_config.num_gpu_blocks, self.sample_on_device_mode)
+        logger.info(f"Warmup run for decode with tokens: {tokens.shape}, start_pos: {start_pos.shape}, page_table: {page_table.shape}")
+        logger.info("Warmup run for decode started")
+        self.model.decode_forward(tokens, start_pos, page_table, kv_cache, trace_mode, sampling_params=sampling_params)
```
[the remark this answers is not in the document]
Comment (skhorasganiTT) [reply vllm/worker/tt_model_runner.py]: @tchedaTT does that sound right?

[the remark this answers is not in the document]
Comment (nostojicTT) [reply vllm/worker/tt_model_runner.py]: I think we can only run one of these two based on the configuration. Models that have sampling on device such as Llama-70B capture only just one trace based on the sampling_params parameter.
https://github.com/tenstorrent/tt-metal/blob/d4af83e47993439631baf80f9ff8356bd17b515f/models/demos/llama3_70b_galaxy/tt/generator.py#L573
https://github.com/tenstorrent/tt-metal/blob/d4af83e47993439631baf80f9ff8356bd17b515f/models/demos/llama3_70b_galaxy/tt/generator.py#L407
So if we pass once anything that will make sampling_params != None, we will record that trace and reuse every next time

File: vllm/worker/tt_model_runner.py [hunk 8]
```diff
@@ -52,6 +52,95 @@
...
+        
+        # 2. Non-greedy sampling (one representative example with penalties)
+        sampling_configs.append(
+            TTSamplingParams(
+                temperature=1.0,
+                top_k=10,
+                top_p=0.9,
+                presence_penalty=PADDING_PRESENCE_PENALTY,
+                frequency_penalty=PADDING_FREQUENCY_PENALTY,
+                repetition_penalty=PADDING_REPETITION_PENALTY
```
Comment (sraizada-tt) [inline vllm/worker/tt_model_runner.py]: this won't trigger penalties trace because default values mean no penalties

Comment (nostojicTT) [reply vllm/worker/tt_model_runner.py]: Didn't know this, will put random numbers

Comment (sraizada-tt) [reply vllm/worker/tt_model_runner.py]: also won't trigger penalties for all sub_meshes in DP -- send something like this:
        sampling_configs.append(
            TTSamplingParams(
                temperature=[1.0]*max_batch_size,
                top_k=[10]*max_batch_size,
                top_p=[0.9]*max_batch_size,
                presence_penalty=[1.2]*max_batch_size,
                frequency_penalty=[1.2]*max_batch_size,
                repetition_penalty=[1.5]*max_batch_size
            )

File: vllm/v1/worker/tt_model_runner.py [hunk 9]
```diff
@@ -83,14 +85,16 @@ def __init__(
...
         # Whether to sample on device
         self.sample_on_device_mode = TTPlatform.sample_on_device_mode
 
         logger.info(
-            "TTModelRunner: trace_mode=%s, sample_on_device_mode=%s",
+            "TTModelRunner: trace_mode=%s, "
+            "sample_on_device_mode=%s, enable_model_warmup=%s",
             self.trace_mode,
             self.sample_on_device_mode,
+            self.enable_model_warmup,
```
Comment (skhorasganiTT) [inline vllm/v1/worker/tt_model_runner.py]: this attribute is no longer needed by the model runner, it can be removed & the log can be done in the worker.
